Add initial code

dnohales opened the pull request:
This add the initial code to automatically import the selected content from Kolibri and create the Flatpak manifest for it.

The README.md file explains how it works.

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,157 @@
+                    help='The Flatpak branch')
+    ap.add_argument('--install-kolibri-flatpak', default=False, action='store_true',
+                    help='If set, automatically install org.learningequality.Kolibri '
+                         'from Falthub')
```
```suggestion
                         'from Flathub')
```

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,157 @@
+MANIFEST_TEMPLATE_PATH = '{}/org.learningequality.Kolibri.Content.json.in'.format(BASE_PATH)
+KOLIBRI_HOME_PATH = '{}/kolibri_home'.format(BASE_PATH)
+CONTENT_JSON_PATH = '{}/content/content.json'.format(KOLIBRI_HOME_PATH)
+_KOLIBRI_FLATPAK_REF = None
```
What's the purpose of this variable which is never assigned to?

dnohales replied:
It's used as cache in `get_kolibri_flatpak_ref`

wjt replied:
No it's not, that function doesn't assign to it.

dnohales replied:
Now yeah, updated.

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,157 @@
+                    help='The ID of the package, it is used as the extension '
+                         'Flatpak ID last segment')
+    ap.add_argument('--branch', default='1.0',
+                    help='The Flatpak branch')
```
This is supposed to match the branch specified in the extension point on the main Flatpak, right? So perhaps:

```suggestion
                    help='The branch specified in the extension point definition')
```

(or better wording)

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,157 @@
+    ap.add_argument('--branch', default='1.0',
+                    help='The Flatpak branch')
+    ap.add_argument('--install-kolibri-flatpak', default=False, action='store_true',
+                    help='If set, automatically install org.learningequality.Kolibri '
```
```suggestion
                    help='automatically install org.learningequality.Kolibri '
```

wjt commented on org.learningequality.Kolibri.Content.json.in:
```diff
@@ -0,0 +1,25 @@
+  "runtime": "org.learningequality.Kolibri",
+  "runtime-version": "master",
+  "sdk": "org.freedesktop.Sdk//19.08",
+  "appstream-compose": false,
```
Would having some appdata mean it is displayed in GNOME Software?

dnohales replied:
I'm really not sure how that part should work, probably yes, we need an AppData template to fill out.

dylanmccall commented on org.learningequality.Kolibri.Content.json.in:
```diff
@@ -0,0 +1,25 @@
+      "name": "kolibri-content",
+      "buildsystem": "simple",
+      "build-commands": [
+        "find kolibri_home/content -type f -exec sh -c 'install -D \"$0\" ${FLATPAK_DEST}/${0#kolibri_home/content/}' {} \\;"
```
This one got lost from the old PR. The Kolibri flatpak expects an extension to contain a "content" subdirectory, so we shouldn't strip off "/content" when installing these files:

```suggestion
        "find kolibri_home/content -type f -exec sh -c 'install -D \"$0\" ${FLATPAK_DEST}/${0#kolibri_home/}' {} \\;"
```

indirectlylit commented:
Minor note: we conventionally run [`black`](https://black.readthedocs.io/en/stable/) on all Python code in LE repos.

Not a hard requirement, but it would minimize diff churn to do that from the start rather than applying later.

dnohales commented:
Updated:

* Used Black with default config to format code
* Created the file `extensions-catalog.json` with the list of extensions to build by default (everything is overridable with CLI options).
* Added AppStream support
* Added a script `list-extensions-catalog` to list the names of the available extensions names in the catalog, useful for build pipelines.

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,397 @@
+KOLIBRI_CONTENT_EXTENSION_BASE_ID = "org.learningequality.Kolibri.Content"
+MANIFEST_TEMPLATE_PATH = "{}/manifest.json.in".format(BASE_PATH)
+MANIFEST = None
+KOLIBRI_HOME_PATH = "{}/kolibri_home".format(BASE_PATH)
```
```suggestion
KOLIBRI_HOME_PATH = os.path.join(BASE_PATH, "kolibri_home")
```

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,397 @@
+
+BASE_PATH = os.path.realpath(os.path.dirname(__file__))
+KOLIBRI_CONTENT_EXTENSION_BASE_ID = "org.learningequality.Kolibri.Content"
+MANIFEST_TEMPLATE_PATH = "{}/manifest.json.in".format(BASE_PATH)
```
```suggestion
MANIFEST_TEMPLATE_PATH = os.path.join(BASE_PATH, "manifest.json.in")
```

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,397 @@
+MANIFEST_TEMPLATE_PATH = "{}/manifest.json.in".format(BASE_PATH)
+MANIFEST = None
+KOLIBRI_HOME_PATH = "{}/kolibri_home".format(BASE_PATH)
+CONTENT_JSON_PATH = "{}/content/content.json".format(KOLIBRI_HOME_PATH)
```
```suggestion
CONTENT_JSON_PATH = os.path.join(KOLIBRI_HOME_PATH, "content/content.json")
```

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,397 @@
+    "metadata_license": "CC0-1.0",
+    "url_homepage": "https://learningequality.org/",
+    "project_group": "Learning Equality",
+    "project_license": "MIT",
```
The code that generates these extensions may be MIT licensed but the content within them is not. I wonder whether it would be better to omit `project_license` by default, or require it to be specified for the specific content? I guess it may be rather hard to extract it from the content itself.

wjt replied:
https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-project_license says:

>  Although the project_license tag is not mandatory, it is highly recommended to include it.

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,397 @@
+    for flag_name in flags_to_use:
+        flag = "--{}".format(flag_name)
+        flag_value = channel.get(flag_name)
+        if type(flag_value) == bool:
```
```suggestion
        if isinstance(flag_value, bool):
```

wjt commented on manifest.json.in:
```diff
@@ -0,0 +1,27 @@
+  "branch": "@BRANCH@",
+  "build-extension": true,
+  "runtime": "org.learningequality.Kolibri",
+  "runtime-version": "master",
```
Kolibri on Flathub is branch `stable`, not `master`. This also ends up being used by the generator when running Kolibri. So I don't think this will work unless you have a hand-built Kolibri Flatpak installed.

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,401 @@
+        + origin_args
+    )
+
+    if len(channel.get("node_ids", [])) > 0:
```
I don't think this is right. I shouldn't be required to list all node IDs in the case where I want to import all content for a channel.

I think this is why when I tried to create an extension, it ended up not having any of the content from that channel in it.

```
$ ./generate-manifest --id Sinugun_samod_hujos --channel channel_id=d69e725e8aea4d19ad58d41c2b4dc315
INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO     Invoking command importchannel network d69e725e8aea4d19ad58d41c2b4dc315
INFO     Downloading data for channel id d69e725e8aea4d19ad58d41c2b4dc315
INFO     Importing ContentTag data
INFO     Importing ContentNode_has_prerequisite data
INFO     Importing ContentNode_related data
INFO     Importing ContentNode_tags data
INFO     Importing ContentNode data
INFO     Importing Language data
INFO     Importing File data
INFO     Importing LocalFile data
INFO     Importing AssessmentMetaData data
INFO     Importing ChannelMetadata data
INFO     Channel d69e725e8aea4d19ad58d41c2b4dc315 successfully imported into the database

wjt@camille:~/src/flathub/org.learningequality.Kolibri$ flatpak run --command=bash app/org.learningequality.Kolibri/x86_64/master
[📦 org.learningequality.Kolibri ~]$ find /app/share/kolibri-content/
/app/share/kolibri-content/
/app/share/kolibri-content/Sinugun_samod_hujos
/app/share/kolibri-content/Sinugun_samod_hujos/.ref
/app/share/kolibri-content/Sinugun_samod_hujos/manifest.json
/app/share/kolibri-content/Sinugun_samod_hujos/content
/app/share/kolibri-content/Sinugun_samod_hujos/content/databases
/app/share/kolibri-content/Sinugun_samod_hujos/content/databases/d69e725e8aea4d19ad58d41c2b4dc315.sqlite3
/app/share/kolibri-content/Sinugun_samod_hujos/content/content.json
/app/share/kolibri-content/Sinugun_samod_hujos/share
/app/share/kolibri-content/Sinugun_samod_hujos/share/metainfo
/app/share/kolibri-content/Sinugun_samod_hujos/share/app-info
/app/share/kolibri-content/Sinugun_samod_hujos/share/app-info/xmls
/app/share/kolibri-content/Sinugun_samod_hujos/share/app-info/xmls/org.learningequality.Kolibri.Content.Sinugun_samod_hujos.xml.gz
/app/share/kolibri-content/Sinugun_samod_hujos/share/appdata
/app/share/kolibri-content/Sinugun_samod_hujos/share/appdata/org.learningequality.Kolibri.Content.Sinugun_samod_hujos.appdata.xml
```

And although there are messages about it from Kolibri in the journal, no channels appear in the interface.

wjt replied:
Yep, if I manually specify a node ID it works.

dnohales replied:
I'm not sure why I had the memory of thinking that not specifying `node_ids` wouldn't download anything, I tried it and it works, gonna fix it.

dnohales replied:
Ok, I'm not crazy, content is not being imported when not using `node_ids` AND after deleting `kolibri_home/content`, looks like you also have to run `deletechannel` and `deletecontent` to really mark the content as deleted. I think I'm gonna do it with `exportchannel` and `exportcontent`, it's the ideal and it's not hard to do.

dnohales commented:
Updated.

wjt commented on generate-manifest:
```diff
@@ -0,0 +1,401 @@
+        action="append",
+        default=[],
+        help="specification of the channel to import, the format would be "
+        "channel_id=<channel ID>[;<parameter_key>=<parameter_value>;...]. parameter_key "
```
This is a cruel command-line format because `;` is the shell statement separator, and `node_ids=blahblah` is a perfectly valid shell command, so I went on a wild goose chase trying to understand why `node_ids` was being ignored!

wjt replied:
Perhaps at least worth a note in the help. I think I'm pretty good at Unix but it really stumped me.

dnohales replied:
I didn't want to create my own format, the really sad thing is that the alternative to this is to use `&` :)

I suppose I can make it to use `:` as separator instead and a string replace.

dnohales commented:
Updated:
* Fixed the channel/content rating dangerous CLI format.
* Now, instead of importing the channel and copy the `kolibri_home/content` folder, I make an import and then an export to another directory. This allows us to reuse the `kolibri_home` directory to speed up incremental builds (I'm not sure if Jenkins is able to reuse the workspace dir).
* If `node_ids` is not specified for a channel, all the content is downloaded.
* Added `no_storage` and `no_database` flags in channels. Since I've noted that Khan Academy SQLite file is 173MB I decided that it's a good idea to allow to create content extensions only with content or only with database files, otherwise, all the Khan Academy extensions should include the database file, adding repeated 173MB to each extension. The problem is if I have to reflect that in the `content.json` file somehow.

dylanmccall commented:
I like that change using exportcontent, and the renamed json files :+1: Looks much tidier!

I don't think there's a need to reflect those flags in `content.json`. The content scanning code in org.learningequality.Kolibri shouldn't need to do anything special in that case. With that said, I'd be cautious about doing database-less or content-less extensions at all until we find ourselves wanting to ship multiple content extensions for the same channel. I'm concerned we'd run into some quirky behaviour trying to manage these, since there isn't a straightforward way to have extensions depend on other extensions. Also, if I recall correctly there's some discussion about having something like `content.json` in Kolibri itself (for the purpose of content exports), so ideally at some point we won't need to generate this file here at all.

wjt commented:
I agree that we shouldn't try too hard to support shipping multiple extensions for the same channel, at least for a first iteration.
